In [1]:
import pandas as pd
import numpy as np
from sklearn.tree import DecisionTreeClassifier
from sklearn.neighbors import KNeighborsClassifier
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score,confusion_matrix
from sklearn.preprocessing import LabelEncoder

In [3]:
#Q1: Load the dataset and display the first five records.

import pandas as pd
df= pd.read_csv("Dataset 2.csv")
df.head()

,UserID,Age,Gender,SubscriptionType,WatchHoursPerWeek,DevicesUsed,FavoriteGenre,AdClicks,MonthlySpend,SubscriptionRenewed
0,1001,22,Female,Basic,23,5,Comedy,13,353,No
1,1002,55,Male,Basic,9,4,Drama,14,317,Yes
2,1003,49,Male,Basic,8,3,Comedy,16,309,No
3,1004,39,Female,Premium,19,5,Drama,45,833,Yes
4,1005,38,Female,Premium,23,5,Sci-Fi,24,804,Yes


In [6]:
#Q2: Determine the number of rows and columns in the dataset

print('shape:',df.shape)
print('\nColumns:')

#Q3 : display all the columns name.
print(df.columns)


shape: (750, 10)

Columns:
Index(['UserID', 'Age', 'Gender', 'SubscriptionType', 'WatchHoursPerWeek',
       'DevicesUsed', 'FavoriteGenre', 'AdClicks', 'MonthlySpend',
       'SubscriptionRenewed'],
      dtype='object')


In [32]:
#Q4: Identify numerical and categorical features.

print('\nData Types')
print(df.dtypes)


Data Types
UserID                 int64
Age                    int64
Gender                 int64
SubscriptionType       int64
WatchHoursPerWeek      int64
DevicesUsed            int64
FavoriteGenre          int64
AdClicks               int64
MonthlySpend           int64
SubscriptionRenewed    int64
dtype: object


In [8]:
df.describe()

,UserID,Age,WatchHoursPerWeek,DevicesUsed,AdClicks,MonthlySpend
count,750.00000,750.000000,750.000000,750.000000,750.000000,750.000000
mean,1375.50000,41.824000,14.236000,3.013333,23.172000,689.905333
std,216.65064,13.739653,7.590243,1.445894,14.829283,371.965835
min,1001.00000,18.000000,1.000000,1.000000,0.000000,199.000000
25%,1188.25000,30.000000,9.000000,2.000000,10.000000,372.250000
50%,1375.50000,42.000000,15.000000,3.000000,23.000000,697.500000
75%,1562.75000,54.000000,19.000000,4.000000,36.000000,821.000000
max,1750.00000,65.000000,38.000000,5.000000,50.000000,1499.000000


In [10]:
#Q5 : Check whether the dataset contains missing values.

df.isnull().sum()

,0
UserID,0
Age,0
Gender,0
SubscriptionType,0
WatchHoursPerWeek,0
DevicesUsed,0
FavoriteGenre,0
AdClicks,0
MonthlySpend,0
SubscriptionRenewed,0


In [33]:
#Q6:Calculate the average age of users.

df["Age"].mean()


np.float64(41.824)

In [12]:
#Q7:Determine the average watch hours per week

df["WatchHoursPerWeek"].mean()


np.float64(14.236)

In [13]:
#Q8 : Find the average monthly spending of users.

df["MonthlySpend"].mean()

np.float64(689.9053333333334)

In [14]:
#Q9: Count the number of users in each subscription category.

df["SubscriptionType"].value_counts()

,count
SubscriptionType,
Basic,342
Premium,279
VIP,129


In [15]:
df["SubscriptionRenewed"].value_counts(normalize=True)*100

,proportion
SubscriptionRenewed,
No,53.733333
Yes,46.266667


In [18]:
# Q11: Convert categorical features into numerical form

from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

df["Gender"] = le.fit_transform(df["Gender"])
df["SubscriptionType"] = le.fit_transform(df["SubscriptionType"])
df["FavoriteGenre"] = le.fit_transform(df["FavoriteGenre"])
df["SubscriptionRenewed"] = le.fit_transform(df["SubscriptionRenewed"])

print(df.head())

   UserID  Age  Gender  SubscriptionType  WatchHoursPerWeek  DevicesUsed  \
0    1001   22       0                 0                 23            5   
1    1002   55       1                 0                  9            4   
2    1003   49       1                 0                  8            3   
3    1004   39       0                 1                 19            5   
4    1005   38       0                 1                 23            5   

   FavoriteGenre  AdClicks  MonthlySpend  SubscriptionRenewed  
0              1        13           353                    0  
1              2        14           317                    1  
2              1        16           309                    0  
3              2        45           833                    1  
4              5        24           804                    1  


In [21]:
# Q12: Define feature set (X) and target variable (y)

X = df.drop("SubscriptionRenewed", axis=1)
y = df["SubscriptionRenewed"]

In [22]:
# Q13: Split the dataset into training and testing sets

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42
)

In [25]:
# Q14: Train a Decision Tree model

from sklearn.tree import DecisionTreeClassifier

dt_model = DecisionTreeClassifier(random_state=42)
dt_model.fit(X_train, y_train)

DecisionTreeClassifier(random_state=42)

In [26]:
# Q15: Evaluate the model using accuracy

from sklearn.metrics import accuracy_score

y_pred_dt = dt_model.predict(X_test)

accuracy_dt = accuracy_score(y_test, y_pred_dt)

print("Decision Tree Accuracy:", accuracy_dt)

Decision Tree Accuracy: 0.5666666666666667


In [27]:

# Q16: Generate and interpret the confusion matrix

from sklearn.metrics import confusion_matrix

cm = confusion_matrix(y_test, y_pred_dt)

print(cm)

[[50 32]
 [33 35]]


In [28]:
# Q17: Train a KNN classifier with K = 5

from sklearn.neighbors import KNeighborsClassifier

knn_model = KNeighborsClassifier(n_neighbors=5)
knn_model.fit(X_train, y_train)

y_pred_knn = knn_model.predict(X_test)

In [29]:
# Q18: Compare the accuracy of KNN with the Decision Tree model

accuracy_knn = accuracy_score(y_test, y_pred_knn)

print("Decision Tree Accuracy:", accuracy_dt)
print("KNN Accuracy:", accuracy_knn)

Decision Tree Accuracy: 0.5666666666666667
KNN Accuracy: 0.6


In [30]:
# Q19: Train a Linear Regression model to predict monthly spending

from sklearn.linear_model import LinearRegression

X_reg = df.drop("MonthlySpend", axis=1)
y_reg = df["MonthlySpend"]

X_train_reg, X_test_reg, y_train_reg, y_test_reg = train_test_split(
    X_reg, y_reg, test_size=0.2, random_state=42
)

lr_model = LinearRegression()
lr_model.fit(X_train_reg, y_train_reg)

LinearRegression()

In [31]:
# Q20: Predict the monthly spending for a new user

new_user = [X_reg.iloc[0].values]

prediction = lr_model.predict(new_user)

print("Predicted Monthly Spend:", prediction[0])

Predicted Monthly Spend: 391.3851628098203


/usr/local/lib/python3.12/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


**BUISNESS REFLECTION ANSWERS**

---


1. Which factors appear to influence subscription renewal the most?

The factors that appear to influence subscription renewal the most are Watch Hours Per Week, Subscription Type, Monthly Spending, and Devices Used. Users who spend more time on the platform and use premium subscription plans are generally more likely to renew their subscriptions.


2. Why is subscription renewal a classification problem?

Answer:
Subscription renewal is a classification problem because the target variable has two categories: Yes or No. The model predicts which class a user belongs to.


3. Why is monthly spending a regression problem?

Answer:
Monthly spending is a regression problem because it is a continuous numerical value. The model predicts an exact amount rather than a category.


4. Which algorithm performed better for renewal prediction?

Answer:
The K-Nearest Neighbors (KNN) algorithm performed better than the Decision Tree model because it achieved higher accuracy on the test dataset. *KNN performed better for renewal prediction with an accuracy of 60%, while the Decision Tree achieved an accuracy of 56.67%.*


5. How could the platform use these predictions to improve customer retention?

Answer:
The platform can identify users who are less likely to renew and provide them with personalized recommendations, discounts, special offers, or engagement campaigns to increase customer retention and reduce churn.